### ISSUE TITLE
`unused-parameter` rule should not trigger on parameters used for their type

### ISSUE THREAD

**Issue by sindresorhus** [realm/SwiftLint#5742]
### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

It should not report for cases where the parameter is used for its type and not value.

```swift
extension Data {
	func jsonDecoded<T: Decodable>(ofType type: T.Type) throws -> T {
		try JSONDecoder().decode(T.self, from: self)
	}
}
```

Here the `type` parameter decides the return type.

### Environment

* SwiftLint version: 0.56.0
* Xcode version: Xcode 15.4, Build version 15F31d
* Installation method used: Installer
* Configuration file:

```yml
only_rules:
  - unused_parameter
```

**Comment by SimplyDanny** [comment]
The rule suggests to write it like:
```swift
extension Data {
  func jsonDecoded<T: Decodable>(ofType _: T.Type) throws -> T {
    try JSONDecoder().decode(T.self, from: self)
  }
}
```
The intention is not to remove the parameter entirely if that's infeasible.

**Comment by sindresorhus** [comment]
My argument is that the parameter *is* technically used when used for its type, not its value. My thinking is that, the parameter is used if removing the parameter makes the function no longer compile.
